### PR TITLE
CONTRIBUTING.md: move content from DocuWiki

### PR DESCRIPTION
**Pull request by Be-ing**
This content was at
https://mixxx.org/wiki/doku.php/contribution_guidelines
but not many developers saw it.

The conventional place for such documents is a CONTRIBUTING.md
file in the Git repository.

### CODE REVIEW

**Comment by Holzhaus** [comment]
Thank you. However I think the CONTRIBUTING.md should be short and concise, because nobody will read it otherwise.

We should add a Quickstart or 'Golden Rules' section at the top with a list of bullet points, and move most of the detailed info to the Wiki or another document.

**Comment by Be-ing** [comment]
I disagree. Take a look at [Rust's CONTRIBUTING.md](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for example.

File: CONTRIBUTING.md
Comment (xerus2000) [inline CONTRIBUTING.md:4]: I think this should be a heading

File: CONTRIBUTING.md
Comment (xerus2000) [inline CONTRIBUTING.md:9]: The numbers here don't make much sense to me

File: CONTRIBUTING.md
Comment (xerus2000) [inline CONTRIBUTING.md:20]:
```suggestion
All of these are automatically built and deployed by our [Jenkins build servers](https://builds.renegadetech.mixxx.org/) whenever a change is committed. You are welcome to open a pull request in any of these repositories.
```

File: CONTRIBUTING.md
Comment (xerus2000) [inline CONTRIBUTING.md:64]: Please divide this into sections, it is rather unwieldy if you are searching for something specific.

E.g. this section could be called "Committing"

File: CONTRIBUTING.md
Comment (xerus2000) [inline CONTRIBUTING.md:71]:
```suggestion
* Please help review other people's pull requests. The continued progress of Mixxx depends on all of us working together. Even if you are not familiar with the area of the code being changed in a pull request, you can help by building the code, verifying that it works as described and giving feedback about the user experience.
```
I don't like that notion of forced reciprocity here.

File: CONTRIBUTING.md
Comment (xerus2000) [inline CONTRIBUTING.md:72]:
```suggestion

If you demonstrate good coding skills, help review pull requests, contribute major features, and show a commitment to Mixxx over time, we may invite you to the core team.
```

I think this should be more of a conclusory/bridging paragraph than an extra point

File: CONTRIBUTING.md
Comment (xerus2000) [inline CONTRIBUTING.md:78]: You can enforce that in the repo settings, no need for an extra statement here.

File: CONTRIBUTING.md
Comment (xerus2000) [inline CONTRIBUTING.md:84]: Section: Merging Pull Requests

File: CONTRIBUTING.md
Comment (xerus2000) [inline CONTRIBUTING.md:80]: Section: Contributing

File: CONTRIBUTING.md
Comment (xerus2000) [inline CONTRIBUTING.md:71]: Section: Pull Requests

**Review by xeruf — CHANGES_REQUESTED** [review]

**Comment by xerus2000** [comment]
I vouch for a section for how to handle review comments, as I did in my collaboration convention: https://xerus2000.github.io/kull/#3-reviewing

I am regularly irritated if someone resolves my non-trivial review comment, because imho I should decide whether the person understood me correctly and that comment it actually solved.

File: CONTRIBUTING.md
Comment (Be-ing) [reply CONTRIBUTING.md:9]: Are you suggesting to change this to an unordered list? Tables of content are usually ordered lists.

File: CONTRIBUTING.md
Comment (Be-ing) [reply CONTRIBUTING.md:78]: Oh, good to know. I think only @rryan has permission to change the repo settings.

**Comment by Be-ing** [comment]
> I am regularly irritated if someone resolves my non-trivial review comment, because imho I should decide whether the person understood me correctly and that comment it actually solved.

Okay, then we should also add something saying that reviewers should resolve comments.

**Comment by xerus2000** [comment]
This is the workflow I use and like:
- reviewer adds comment
- if it is a trivial issue such as a typo, the contributor can fix the issue and resolve the comment
- if the contributor is unsure what is meant, he asks for more information
- otherwise, he may fix it (according to his understanding) and comment with a link to the change (commit/updated file) and the reviewer will resolve the comment if he deems that fix adequate or leave another comment

You can have a look at the (unfortunately german) resolved review comments here illustrating that: https://github.com/CAU-Kiel-Tech-Inf/socha/pull/260

Additional rules that seem sensible:
- if the original reviewer does not get back to a commented fix within a set period (maybe a week), the PR owner may resolve the conversation
- a review comment may contain a task to be addressed later, then it can be added to a checklist/issue and resolved with a reference to that

The important thing is that the contributor refrains from closing non-trivial issues. If I can't rely on that, I regularly look at "resolved" conversations regularly to check whether they are actually resolved.

**Comment by daschuer** [comment]
I would prefere that only the reviewer resolved the conversations.

There is already a kind of "resolved" state for the contributor, because the conversation is becoming outdated and it is removed from the diff view.

**Comment by Holzhaus** [comment]
I'm actually in the other camp. When I look at a PR and see that some if my previous comments have not been resolved yet I assume that the contributer still has stuff to do and the PR is not ready to re-review.

Also, if we say that only the original reviewer may resolve issues this will essentially mean that we need to wait until every reviewer has reviewed the PR a second time. That will probably prolong our review times even more.

**Comment by daschuer** [comment]
> Also, if we say that only the original reviewer may resolve issues this will essentially mean that we need to wait until every reviewer has reviewed the PR a second time. That will probably prolong our review times even more.

This is already the case without the resolve feature. Review, fail, review again approve.
(In the ideal case the reviewer should also verify the PR by a manual test before merge.)

I can live with the contributor resolve the comments, however this is an extra click during the second review, and it becomes worse if a third review is required or a second reviewer has done some work.

Du to the collapsed state you cannot instantly see if it is an old finished comment or a new just fixed comment.
And who has commented.

The reviewers needs to open and close many resolved comments to check them.

**Comment by Holzhaus** [comment]
> This is already the case without the resolve feature. Review, fail, review again approve.
(In the ideal case the reviewer should also verify the PR by a manual test before merge.)

I didn't realize we have this requirement. Before I think it was possible to have person A open a PR, person B leaves some review comments, person A makes changes, person C approves and merges.

With this system we need to wait for person B show review again, too, before we can merge.

**Comment by daschuer** [comment]
Yes, this was my understanding.
Only the original author can approve a previous failed review.

**Comment by xerus2000** [comment]
> > This is already the case without the resolve feature. Review, fail, review again approve.
> > (In the ideal case the reviewer should also verify the PR by a manual test before merge.)
> 
> I didn't realize we have this requirement. Before I think it was possible to have person A open a PR, person B leaves some review comments, person A makes changes, person C approves and merges.

From the PRs I worked on so far, almost always the original reviewer re-reviewed when I made changes - since review comments can be misunderstood/misinterpreted, I think this is important.

File: CONTRIBUTING.md
Comment (xerus2000) [reply CONTRIBUTING.md:9]: On github mostly not, from what I've seen, for example:
https://github.com/sghpjuikit/player/blob/master/README.md
https://github.com/sharkdp/bat/blob/master/README.md

So, yes, I would change this to an unordered list, and maybe even remove the heading "Table of Contents" - it is self-evident.

**Comment by xerus2000** [comment]
But I don't think the original reviewer should be required to grant explicit approval, i.e. perform a second full review. If all his comments are resolved (at least code-wise) and there is another approval, it should be fine to merge.

**Comment by daschuer** [comment]
There is no IMHO no question about this.
The GitHub workflow works like that:
Due to the failed review there is a red check mark. Before merge, it should be turned into green. This is the responsibility of the original reviewer. How much he trust the contributor, is up to him.

**Comment by Holzhaus** [comment]
That means we should always use the review mechanism instead of simple comments. If you leave comments that don't necessarily need to be taken care of before merge, you can use "Comment" instead of "approve" or "Request changes". And if you requested changes that are resolved now, you need to explicitly "Approve" them using the review mechanism, not just comment "LGTM" (because else the red cross will persist).

I'm okay with that. What about pending reviewers though? There will be a yellow dot if a review was requested but not done. Should we always wait until everyone in the reviewers list approved or commented? If someone contributor requested review by someone but that person does not replay or review (but someone else does), I'd be okay with merging anyway to speed up the process but it might also make sense to wait so that we can use the reviewers list to know if a PR is ready to merge at a glance.

**Comment by daschuer** [comment]
We have the rule that a single LGTM (old) or an "Approved" (new) review is sufficient. 
I normally write a LGTM comment as well, mainly to thank the contributor and not have just the anonymous green check mark.

Since the reviewer does the merge, we have another safety net.

So far, the review request is a kind of ping for review, to me and does not affect the rule above.

If a review from one person is desired, we should write it as a comment as well.

**Comment by Holzhaus** [comment]
I still think we should add a *Quickstart* section at the top, outlining the most important points for drive-by contributors, each linking to the corresponding verbose section e.g.:

- Be polite and respectful to each otther (link to COC)
- Work in feature branches, an make sure to base your work on the appropriate base branch
- Use our pre-commit hooks
- Open a draft PR to get early feedback
- ...

**Review by daschuer — APPROVED** [review]
Thank you. LGTM.

**Comment by daschuer** [comment]
@Holzhaus: merge?

**Comment by Holzhaus** [comment]
I still disagree strongly with the wall of text that hides important information. It's too much too to require contributors to read that much IMHO. But it's certainly better than the current situation, so let's merge.

**Comment by xeruf** [comment]
Um, now this was merged without addressing all the review comments...

**Comment by Holzhaus** [comment]
Yes. I'm not happy with CONTRIBUTING.md either, let's do that in other PRs.

**Pull request merged**
